### PR TITLE
Wip in radius

### PR DESCRIPTION
**Pull request by brotatotes**
Speed optimizations to In-Radius and In-Cone

### CODE REVIEW

File: netlogo-core/src/main/agent/InRadiusOrCone.scala [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+      }
+    }
+
+    result
+  }
+
+  // helper method for inCone()
+  // for each world W at offsets offsetX and offsetY, check if any point
+  // in W is within radius by checking W's closest point
+  private def closestPointIsInRadius(x: Double, y: Double, offsetX: Int, offsetY: Int, r: Double): Boolean = {
```
Comment (mrerrormessage) [inline netlogo-core/src/main/agent/InRadiusOrCone.scala]: This method can be made `@scala.inline`.

File: netlogo-core/src/main/agent/InRadiusOrCone.scala [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+      closestX = x
+    }
+
+    world.protractor.distance(closestX, closestY, x, y, false) <= r
+  }
+
+  // helper method for inCone().
+  // check if (x, y) is in the cone with center (cx, cy) , radius r, half-angle half, and central
+  // line of the cone having heading h.
+  private def isInCone(x: Double, y: Double, cx: Double, cy: Double, r: Double, half: Double, h: Double): Boolean = {
```
Comment (mrerrormessage) [inline netlogo-core/src/main/agent/InRadiusOrCone.scala]: This method can be made `@scala.inline` here and on all `private` methods in this class.

File: test/commands/InRadius.txt [hunk 3]
```diff
@@ -1,315 +1,460 @@
...
 InRadiusLargeRadius4_2D
   O> resize-world -10 0 -10 0
+  O> create-turtles 10
   O> __change-topology true true
   [ count patches in-radius 14 ] of patch 0 0 => 121
   [ count patches in-radius 14 ] of patch -3 0 => 121
   [ count patches in-radius 14 ] of patch max-pxcor 0 => 121
   [ count patches in-radius 14 ] of patch 0 min-pycor => 121
+  [ count turtles in-radius 1 ] of patch -10 -10 => 0
+  [ count turtles in-radius 1 ] of patch max-pxcor max-pycor => 10
```
Comment (mrerrormessage) [inline test/commands/InRadius.txt]: Our code special-cases radius 2, but we have no test cases which exercise that. Let's add some.

File: netlogo-core/src/main/agent/Topology.scala [hunk 4]
```diff
@@ -181,6 +188,122 @@ extends Neighbors {
...
+    } else { // !high_within
+      if (xWraps) {
+        mergeAdd((offset + 0, offset + x + r - world.worldWidth + 1), arr)
+      }
+      mergeAdd((offset + x - r, offset + world.worldWidth), arr)
+    }
+
+  }
+
+  // helper fo getRegion/getRegionRow
```
Comment (mrerrormessage) [inline netlogo-core/src/main/agent/Topology.scala]: I would kill the "helper for" comment here and in the rest of the file.

File: netlogo-core/src/main/agent/InRadiusOrCone.scala [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+      region = regionIterator.next()
+      length = region._2 - region._1
+      System.arraycopy(worldPatches, region._1, patches, curr, length)
+      curr += length
+    }
+
+    end = curr
+  }
+
+  // helper method to create cachedIDs set
```
Comment (mrerrormessage) [inline netlogo-core/src/main/agent/InRadiusOrCone.scala]: Drop helper method comments throughout.

**Comment by brotatotes** [comment]
[Wiki](https://github.com/NetLogo/NetLogo/wiki/Notes-on-in-radius) updated.

**Comment by mrerrormessage** [comment]
I see the major change has been to round x and y. As long as this gives the same numerical results (or same-ish numerical results) in Lennard-Jones, I think it can be merged. There appears to be a problem with rebasing here, as this includes a large number of commits not made on this branch. It might be easiest to take a new branch off master and cherry-pick commits from this work onto that branch, then force push that up to this PR.

**Pull request merged**